### PR TITLE
ARROW-13466: [R] make installation fail if Arrow C++ dependencies cannot be installed

### PR DESCRIPTION
**Pull request by nealrichardson**
The first commit should fail on the "without-arrow" CI build. I'll then add an env var so we can still test that setup.

### CODE REVIEW

**Comment by nealrichardson** [comment]
@github-actions crossbow submit test-r-without-arrow

File: .github/workflows/r-without-arrow.yml [hunk 1]
```diff
@@ -73,7 +73,7 @@ jobs:
         run: |
           sudo sysctl -w kernel.core_pattern="core.%e.%p"
           ulimit -c unlimited
-          archery docker run -e LIBARROW_DOWNLOAD=FALSE -e LIBARROW_BUILD=FALSE -e TEST_R_WITH_ARROW=FALSE -e NOT_CRAN=FALSE r
+          archery docker run -e LIBARROW_DOWNLOAD=FALSE -e LIBARROW_BUILD=FALSE -e TEST_R_WITHOUT_LIBARROW=true -e NOT_CRAN=FALSE r
```
Comment (jonkeane) [inline .github/workflows/r-without-arrow.yml:76]: Very much a nit, but all the others are all-caps `TRUE`/`FALSE`

```suggestion
          archery docker run -e LIBARROW_DOWNLOAD=FALSE -e LIBARROW_BUILD=FALSE -e TEST_R_WITHOUT_LIBARROW=TRUE -e NOT_CRAN=FALSE r
```

File: r/configure [hunk 2]
```diff
@@ -255,17 +254,19 @@ if [ $? -eq 0 ] || [ "$UNAME" = "Darwin" ]; then
...
 else
-  if [ "$UNAME" = "Darwin" ]; then
-    # Just for debugging: is this possible?
-    echo "Test to load header failed. Command:"
-    echo "$TEST_CMD"
-  fi
   echo "------------------------- NOTE ---------------------------"
   echo "See https://arrow.apache.org/docs/r/articles/install.html"
   echo "for help installing Arrow C++ libraries"
   echo "---------------------------------------------------------"
```
Comment (jonkeane) [inline r/configure:260]:
```suggestion
  echo "------------------------- NOTE ---------------------------"
  echo "There was an issue building or finding the Arrow C++ libraries."
  echo "See https://arrow.apache.org/docs/r/articles/install.html"
  echo "for help installing Arrow C++ libraries"
  echo "---------------------------------------------------------"
```

I added a bit to make it a little more clear/obvious what happened there. We have the URL which explains it, but I think stating definitively that there was a problem building/finding is good/makes this more understandable. Open to other wordings of course.

File: r/tests/testthat/test-arrow.R [hunk 3]
```diff
@@ -76,4 +77,4 @@ test_that("MemoryPool calls gc() to free memory when allocation fails (ARROW-100
   # but it should gc() and retry (and fail again)
   expect_error(BufferOutputStream$create(2**60))
   expect_true(env$gc_was_called)
-})
+})
```
Comment (jonkeane) [inline r/tests/testthat/test-arrow.R:80]:
```suggestion
})

```

To appease the lintr, styler _should_ do this, but this is easy enough.

File: r/vignettes/install.Rmd [hunk 4]
```diff
@@ -174,16 +174,19 @@ The intent is that `install.packages("arrow")` will just work and handle all C++
...
+See https://arrow.apache.org/docs/r/articles/install.html
+for help installing Arrow C++ libraries
+---------------------------------------------------------
 ```
 
-for every `arrow` function you call,
-that means that installing the package failed to retrieve or build C++ libraries
+in the output when the package fails to install,
+that means that installation failed to retrieve or build C++ libraries
 compatible with the current version of the R package.
```
Comment (jonkeane) [inline r/vignettes/install.Rmd:190]: A few lines below (though I can't seem to commend on them in the GH UI), we should probably suggest _either_ `arrow::install_arrow(verbose = TRUE)` or `Sys.setenv(ARROW_R_DEV=TRUE); install.packages("arrow")` or however the person was installing it, now that they won't have the shell packages to call `arrow::install_arrow()` from.

**Review by jonkeane — COMMENTED** [review]
This looks great, and I think will very much improve our messaging / the user experience when things go wrong.

File: r/tests/testthat/test-arrow.R (hunk 3 above)
Comment (nealrichardson) [reply r/tests/testthat/test-arrow.R:80]: Yeah this is a case of vscode sometimes stripping the trailing newline

File: ci/scripts/r_test.sh [hunk 5]
```diff
@@ -41,12 +41,12 @@ if [ "$ARROW_R_DEV" = "TRUE" ]; then
     export NOT_CRAN=true
   fi
 fi
-: ${TEST_R_WITH_ARROW:=TRUE}
-export TEST_R_WITH_ARROW=$TEST_R_WITH_ARROW
 
 export _R_CHECK_CRAN_INCOMING_REMOTE_=FALSE
-# --run-donttest was used in R < 4.0, this is used now
-export _R_CHECK_DONTTEST_EXAMPLES_=$TEST_R_WITH_ARROW
+if [ "$TEST_R_WITHOUT_LIBARROW" != "true" ]; then
```
Comment (nealrichardson) [inline ci/scripts/r_test.sh:46]:
```suggestion
if [ "$TEST_R_WITHOUT_LIBARROW" != "TRUE" ]; then
```

File: r/configure [hunk 6]
```diff
@@ -255,17 +254,19 @@ if [ $? -eq 0 ] || [ "$UNAME" = "Darwin" ]; then
...
   PKG_LIBS=""
   PKG_CFLAGS=""
+  if [ "$UNAME" != "SunOS" ] && [ "$TEST_R_WITHOUT_LIBARROW" != "true" ]; then
```
Comment (nealrichardson) [inline r/configure:263]:
```suggestion
  if [ "$UNAME" != "SunOS" ] && [ "$TEST_R_WITHOUT_LIBARROW" != "TRUE" ]; then
```

**Pull request closed without merging**